Скачайте файл с предложениями (sentences.txt).
Каждая строка в файле соответствует одному предложению. Считайте их, приведите каждую к нижнему регистру с помощью строковой функции lower().
Произведите токенизацию, то есть разбиение текстов на слова. Для этого можно воспользоваться регулярным выражением, которое считает разделителем любой символ, не являющийся буквой: re.split('[^a-z]', t). Не забудьте удалить пустые слова после разделения.

In [16]:
from collections import Counter
from scipy.spatial import distance
import re
import numpy as np

In [17]:
lines = []
with open('sentences.txt') as data_file:
    for line in data_file.readlines():
        lines.append([word for word in re.split('[^a-z]',line.strip().lower())])

In [18]:
lines = [[word for word in line if word != ''] for line in lines]
print(len(lines))

22


Составьте список всех слов, встречающихся в предложениях. Сопоставьте каждому слову индекс от нуля до (d - 1), где d — число различных слов в предложениях. Для этого удобно воспользоваться структурой dict.

In [19]:
words = {}
word_counter = 0
for line in lines:
    for word in line:
        if word not in words:
            words[word] = word_counter
            word_counter += 1
print(word_counter)

254


Создайте матрицу размера n * d, где n — число предложений. 

In [20]:
word_matrix = np.zeros((len(lines), len(words)))

Заполните ее: элемент с индексом (i, j) в этой матрице должен быть равен количеству вхождений j-го слова в i-е предложение. У вас должна получиться матрица размера 22 * 254.

In [21]:
for i, line in enumerate(lines):
    counter = Counter(line)
    for j,word in enumerate(words):
        word_matrix[i, j] = counter[word]
print(word_matrix)

[[1. 1. 1. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 [0. 0. 2. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [1. 0. 1. ... 0. 0. 0.]
 [0. 0. 1. ... 1. 1. 1.]]


Найдите косинусное расстояние от предложения в самой первой строке (In comparison to dogs, cats have not undergone...) до всех остальных с помощью функции scipy.spatial.distance.cosine. Какие номера у двух предложений, ближайших к нему по этому расстоянию (строки нумеруются с нуля)? 

In [22]:
cosin_distances = {}
for i,line in enumerate(word_matrix):
    cosin_distances[i] = (distance.cosine(word_matrix[0], line))
{print(k, v) for k, v in sorted(cosin_distances.items(), key=lambda item: item[1])}

0 0.0
6 0.7327387580875756
4 0.7770887149698589
21 0.8250364469440588
10 0.8328165362273942
12 0.8396432548525454
16 0.8406361854220809
20 0.8427572744917122
2 0.8644738145642124
13 0.8703592552895671
14 0.8740118423302576
11 0.8804771390665607
8 0.8842724875284311
19 0.8885443574849294
3 0.8951715163278082
9 0.9055088817476932
7 0.9258750683338899
5 0.9402385695332803
15 0.9442721787424647
18 0.9442721787424647
1 0.9527544408738466
17 0.956644501523794


{None}